torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,440 @@
+    gm.graph.move_after(need_sort_nodes, last_split_reshape_node)
+
+    gm.graph.eliminate_dead_code()
+
+
+def _fuse_with_cat(
+    gm: IterGraphModule,
+    comm_blocks: List[CommBlock],
+    node_indices: Dict[fx.Node, int],
+) -> CommBlock:
+    """
+    Given a list of CommBlock (only allreduce), fuse the CommBlocks using concat.
+    """
+    # Find the last input node.
+    last_input_node = comm_blocks[0].inputs[0]
+    last_input_index = -1
+    all_input_nodes = []
+    for comm_block in comm_blocks:
+        input_node = comm_block.inputs[0]
+        # If the input node is a clone, this is CommTensor based implementation.
+        if input_node.name.startswith("clone"):
+            input_node = cast(fx.Node, input_node.args[0])
+        all_input_nodes.append(input_node)
+        index = node_indices[input_node]
+        if index >= last_input_index:
+            assert index != last_input_index
+            last_input_node = input_node
+            last_input_index = index
+
+    # Flatten all the inputs right after the last input is ready.
+    with gm.graph.inserting_after(last_input_node):
+        cat_inputs = []
+        for input_node in all_input_nodes:
+            cat_inputs.append(
+                _call_function(
+                    gm, fake_tensor_mode, None, aten.flatten.using_ints, input_node
+                )
+            )
+
+    with gm.graph.inserting_after(cat_inputs[0]):
+        cat_node = _call_function(gm, fake_tensor_mode, None, aten.cat, cat_inputs)
+
+    # Create a new Comm node.
+    last_comm = comm_blocks[-1]
+    last_comm_node = last_comm.comm_node
+    last_wait_node = last_comm.wait_nodes[0]
+    with gm.graph.inserting_after(cat_node):
+        flatten_args, spec = tree_flatten((last_comm_node.args, last_comm_node.kwargs))
+        flatten_args[0] = cat_node
+        args, kwargs = tree_unflatten(flatten_args, spec)
+        fused_comm_node = _call_function(
+            gm,
+            fake_tensor_mode,
+            cat_node.meta["val"],
+            last_comm_node.target,
+            *args,
+            **kwargs,
+        )
+
+    # Create a new Wait node.
+    with gm.graph.inserting_after(fused_comm_node):
+        flatten_args, spec = tree_flatten((last_wait_node.args, last_wait_node.kwargs))
+        flatten_args[0] = fused_comm_node
+        args, kwargs = tree_unflatten(flatten_args, spec)
+        fused_wait_node = _call_function(
+            gm,
+            fake_tensor_mode,
+            cat_node.meta["val"],
+            last_wait_node.target,
+            *args,
+            **kwargs,
+        )
+
+    # Move the fused_comm_node and its args to right after the source node
+    nodes_to_move = cat_inputs + [cat_node, fused_comm_node, fused_wait_node]
+    gm.graph.move_after(nodes_to_move, last_input_node)
+
+    tensor_meta = cat_node.meta.get("tensor_meta")
+    fused_comm_block = CommBlock(
+        shape=tensor_meta.shape,  # type: [union-attr]
```

Comment:
When the input graph has symbolic shape enabled, `.shape` will contain `SymInt`s and `comm_fusion_with_cat` will break the graph by using `SymInt`s not defined in the graph as some nodes' arguments.

In the long term we need to support symbolic shape. For now we need to convert the `SymInt`s to `int`s here (e.g. `shape=torch.Size(int(s) for s in torch_meta.shape)`).

Reply:
Applying the change and will add a TODO here.